File: Data Structures & Algorithms/maximum-subarray/submission-4.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        curr, ans = nums[0], nums[0]
        for i in range(1, len(nums)):
            curr = max(curr + nums[i], nums[i])
            ans = max(ans, curr)
        return ans

File: Data Structures & Algorithms/subsets-ii/submission-4.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> curr;
        dfs(0, curr, nums);
        return result;
    }
    void dfs(int i, vector<int> curr, vector<int>& nums) {
        if (i == nums.size()) {
            result.push_back(curr);
            return;
        }
        curr.push_back(nums[i]);
        dfs(i + 1, curr, nums);
        curr.pop_back();
        while (i + 1 < nums.size() and nums[i] == nums[i + 1]) {
            i++;
        }
        dfs(i + 1, curr, nums);
    }
};
